### PR TITLE
Show parsing errors in context of line where they occur

### LINKED ISSUE #15: Better errors, line numbers, dump the ast etc
parser errors in particular lack context

and for other, like using the wrong function name, it's also obscure 

at least should say which is wrong, eg
```
$ typo("abc")
== Parse ==> typo("abc")
== Macro ==> typo("abc")
== Eval  ==> <err: <not a function: ERROR>>
```

### LINKED ISSUE #166: limit number of parsing errors
when parsing errors start specially on a large block (like someone pasting go code into grol bot) should limit errors to first few...

could be a grol setting or I guess handled in discord bot

related to lack of line numbers etc too

### PR DESCRIPTION
**Pull request by ldemailly**
Now:
```
$ if true {
>    len{}
13:04:44.844 [ERR] parser error: expected next token to be `(`, got `{` instead:
   len{}
      ^
$ 1e+999 - 6
13:05:15.404 [ERR] parser error: could not parse "1e+999" as float:
1e+999 - 6
     ^
$ if foo {  % blah
13:07:21.791 [ERR] parser error: no prefix parse function for `%` found:
if foo {  % blah
          ^
```
Fixes #15  (at least for parsing errors, for eval time we may want a stack trace)
Fixes #166  (will follow up in discord bot PR)

### CODE REVIEW

File: eval/eval.go [hunk 1]
```diff
@@ -327,7 +327,7 @@ func evalIndexExpression(left, index object.Object) object.Object {
 		if idx < 0 || idx >= int64(len(str)) {
 			return object.NULL
 		}
-		return object.Integer{Value: int64(str[idx])}
+		return object.Integer{Value: int64(str[idx])} //nolint:gosec // https://github.com/securego/gosec/issues/1185
```
Comment (ccoVeille) [inline eval/eval.go:330]: A comment explaining the problem would be appreciated. Yes, there is a link, and it's great. but someone reading needs tpneedo open the link. People are lazy. I can tell I'm lazy 😬

File: parser/parser.go [hunk 2]
```diff
@@ -273,14 +275,25 @@ func (p *Parser) expectPeek(t token.Type) bool {
 	return false
 }
 
+// ErrorContext returns the current line and a pointer to the error position.
+// If prev is true, the error position is relative to the previous token instead of current one.
+func (p *Parser) ErrorContext(prev bool) string {
```
Comment (ccoVeille) [inline parser/parser.go:280]: Two remarks:
- context is connoted in Go. I would have avoided
- it's not really about error. You do use it to say where the error is, but there is nothing about error here. The function can be called even when there is no error

File: parser/parser.go [hunk 3]
```diff
@@ -273,14 +275,25 @@ func (p *Parser) expectPeek(t token.Type) bool {
...
+	line, errPos := p.l.CurrentLine()
+	if prev {
+		errPos -= (p.l.Pos() - p.prevPos)
```
Comment (ccoVeille) [inline parser/parser.go:283]: This line require a comment as the computing you do is cryptic.

People don't read the function comment.

This line should be documented

File: parser/parser.go (hunk 2 above)
Comment (ccoVeille) [reply parser/parser.go:280]: Maybe tgis method could call Errors() (it's the right object, right ?) and return a different result when there is no error

File: parser/parser_test.go [hunk 4]
```diff
@@ -364,3 +364,18 @@ func TestIncompleteBlockComment(t *testing.T) {
 		}
 	}
 }
+
+func TestNilToken(t *testing.T) {
```
Comment (ccoVeille) [inline parser/parser_test.go:368]: I don't think people will understand what this test is about

File: token/token.go [hunk 5]
```diff
@@ -308,12 +309,16 @@ func LookupIdent(ident string) *Token {
...
-// (even though that's better than string comparaisons).
+// (even though that's better than string comparisons).
 func ByType(t Type) *Token {
 	return tToT[t]
 }
 
 func (t *Token) Literal() string {
+	if t == nil {
+		log.Critf("Nil token .Literal() called")
+		return "NIL_TOKEN"
```
Comment (ccoVeille) [inline token/token.go:320]: Is there a risk if someone write NIL_TOKEN and ask grol to parse it ?

**Comment by ccoVeille** [comment]
I would like to recommend you

https://github.com/meyermarcel/annot

(Hi @meyermarcel 👋)

File: eval/eval.go (hunk 1 above)
Comment (ldemailly) [reply eval/eval.go:330]: strings[idx] are bytes (unsigned uint8) and the linter thinks that 0-255 value can't fit into a int64 which is completely stupid

File: parser/parser.go (hunk 2 above)
Comment (ldemailly) [reply parser/parser.go:280]: it's used to construct the errors so it can't use the errors

File: token/token.go (hunk 5 above)
Comment (ldemailly) [reply token/token.go:320]: no it's if you lookup a token in a map and get nil and then call this method (in this case it was @ being looked up and not found in the map and then still used to print error, I fixed that but left the log for the next time a bug like that shows up (though the natural panic NPE might be fine too)

**Comment by ldemailly** [comment]
> I would like to recommend you
> 
> https://github.com/meyermarcel/annot
> 
> (Hi @meyermarcel 👋)

fancy :) I think I'm ok with my lowkey single ^ though for now :)

File: parser/parser_test.go (hunk 4 above)
Comment (ldemailly) [reply parser/parser_test.go:368]: changing to this in #171 
```go
func TestInvalidToken(t *testing.T) {
	// This turns into the invalid token which in turn can't be found in prefix map,
	// it used to crash in earlier versions, this is the regression for that crash.
	// it also checks the error message in first line + first column case.
	inp := "@"
```

File: parser/parser.go (hunk 3 above)
Comment (ldemailly) [reply parser/parser.go:283]: doing in #171  - note that if you don't read function comment... yet read code/non function comment... seems a bit odd / spurious argument

File: parser/parser_test.go (hunk 4 above)
Comment (ccoVeille) [reply parser/parser_test.go:368]: OK 👍

File: eval/eval.go (hunk 1 above)
Comment (ldemailly) [reply eval/eval.go:330]: the link to that bug is enough. and hopefully it get fixed, though it pisses me off golanglint-ci action pin doesn't pin the linters themselves and stuff start breaking mid PRs </rant>

**Comment by ccoVeille** [comment]
> > I would like to recommend you
> > 
> > https://github.com/meyermarcel/annot
> > 
> > (Hi @meyermarcel 👋)
> 
> fancy :) I think I'm ok with my lowkey single ^ though for now :) 

Noted. I also found https://github.com/caarlos0/mdtree

New and interesting

File: parser/parser.go (hunk 3 above)
Comment (ccoVeille) [reply parser/parser.go:283]: I know 😬

**Pull request merged**